### DISCUSSION TITLE
`Antenna` class

### DISCUSSION THREAD

**Discussion by jhillairet** [scikit-rf/scikit-rf#693] [Ideas]
Talking about `scikit-rf` in a domestic conference, a few people asked me about antenna stuff and supporting methods.

As an antenna could be seen as an N-port Network (generally a 1-port) and `scikit-rf` being the Python swiss knife of RF engineering, it would have indeed some sense to propose things for antennas. An `Antenna` class which could provide additional features than `Network`. The following ideas were suggested (non-exhaustive list!):

* Easy plotting of radiation pattern. Many users use the library for its easy and convenient plotting capabilities, and wish to do the same for radiation patterns
* Far field radiation patterns from near field
* Radiation pattern and anechoic chamber data acquisition 
* Simulation of S-parameters behaviour of typical antennas (say, dipole, etc)

Any other suggestions?

**Comment by arsenovic** [comment]
i dont have experience too much with needs of antenna designers, but would be a nice addition.

**Comment by mhuser** [comment]
There is sometimes a need to "[match this antenna](https://www.johansontechnology.com/downloads/johanson-technology-antenna-mounting-techniques.pdf)", for small ceramic chip antennas on PCB using pi or T networks of capacitors and inductors. This could be done by hand by using the smith chart and an iterative process. The target is to optimize power transfer for small signal power amplifiers.
This is only a random idea: it is more complex than it sounds if considering the chip components have values in normalized series as well as tolerances and non-ideal frequency behaviour.

**Comment by cafeclimber** [comment]
I've been thinking a lot about this and think I'd like to tackle it. Importantly, I think we need to agree on the purpose of the class and what a good interface would be.

As for the purpose of the `Antenna` class, in my opinion it should provide the ability to store full 3D radiation pattern data. This is complex because this is 3D data (mag over theta, phi, freq). @jhillairet mentioned it's basically a 1-port network, which is true in a sense, but I think what we're really looking for is a `Network` for every angle of incidence. 

This leads me to think a 2D list of Network objects would be a good way to represent the data internally. However this also carries a lot of extra baggage. We're not going to have different frequency ranges for each position for example. Would love to hear folks thoughts on alternative. Having data 2D like this would allow us to index by theta/phi for example. (`antenna[theta][phi]` maybe?)

Now as for the interface, beyond what I mentioned above re: indexing, I think the following methods should exist (open to change of course):

- `plot_az_cut(phi: float)`
- `plot_el_cut(theta: float)`
- `plot_3d()`
- `hpbw() #property?`
- `first_null_bw() #property?`
- `dipole()/patch()/etc... #class methods that return an instance of Antenna with simulated data`
- other methods provided by 

@jhillairet also mentioned anechoic chamber data acquisition. Could you expand on this? I think if `Antenna` were implemented in this way, it could greatly benefit (selfishly) my anechoic measurement program, [PyChamber](https://github.com/pychamber/PyChamber), but if there's ways that the `Antenna` class could help users of scikit-rf with chamber measurements, it would be cool to implement that as well!

**Reply by jhillairet** [reply]
> I've been thinking a lot about this and think I'd like to tackle it. Importantly, I think we need to agree on the purpose of the class and what a good interface would be.

Great !
 
> As for the purpose of the `Antenna` class, in my opinion it should provide the ability to store full 3D radiation pattern data. This is complex because this is 3D data (mag over theta, phi, freq). @jhillairet mentioned it's basically a 1-port network, which is true in a sense, but I think what we're really looking for is a `Network` for every angle of incidence.

Yes, I agree, a `Network` for each angle. (because we can image more than one port !)

Instead of creating a new kind of class, this could be a `NetworkSet`, where each `Network` would have named parameters to define the angle (cf bottom of the page https://scikit-rf.readthedocs.io/en/latest/tutorials/NetworkSet.html). Filtering named parameters of a NetworkSet is easy, so one can easily get all the s param vs freq for a given angle.

New method of the Antenna class could be used to get cut planes for a given frequency, etc.
 
> @jhillairet also mentioned anechoic chamber data acquisition. Could you expand on this? I think if `Antenna` were implemented in this way, it could greatly benefit (selfishly) my anechoic measurement program, [PyChamber](https://github.com/pychamber/PyChamber), but if there's ways that the `Antenna` class could help users of scikit-rf with chamber measurements, it would be cool to implement that as well!

It would be great if the output of PyChamber would be such an Antenna class!

**Reply by cafeclimber** [reply]
So internally, PyChamber already does this using a `NetworkSet` :D All data and both polarizations are stored in one `NetworkSet` with parameters used to tag each network properly. I just worry about the overhead. Creating a new `Network` for each measurement seems like a lot! 

What would you think about an ndarray with shape `(len(thetas), len(phis), len(freqs)`? Then in the class or during instantiation have methods that do the indexing properly (i.e. convert degrees to an integer index). Then someone could have an `Antenna` Object with varying resolution or varying extents? Of course, this would preclude having a multi-port antenna like you mention, I think.

Also if we _did_ use `NetworkSet`, I think it (or `Network`, not sure) would need some updates. Because params is a dictionary, these objects are _not pickleable_. Additionally, `NetworkSet.write_spreadsheet()` does not work (I've not filed an issue yet, apologies!)

**Reply by cafeclimber** [reply]
These are the relevant files in PyChamber:

* [network_model.py](https://github.com/PyChamber/PyChamber/blob/main/pychamber/models/ntwk_model.py)
* [experiment.py run_experiment()](https://github.com/PyChamber/PyChamber/blob/main/pychamber/plugins/experiment.py#L409)

**Reply by arsenovic** [reply]
using xarray for labeled nd-arrays (with coordinates) is really handy. might be helpful

**Reply by cafeclimber** [reply]
[This](https://github.com/pydata/xarray) xarray? Looks super neat! Just have to decide if it's worth the introduction of another dependency. Since our "labels" would always be the same, I'm tempted to say no? That would just be implemented in the class. But if there's other functionality you think would be useful, lmk! I'll have a deeper look at it

**Reply by arsenovic** [reply]
yes thats it. but i agree with you. if networkset used this internally, then antenna could just subclass it . would require more work obviously. 
really, network should use this, and networkset should not exist, but hey thats scikit-rf 2.0.

**Reply by cafeclimber** [reply]
How would you see `Network` taking over the functionality of `NetworkSet`?

**Reply by jhillairet** [reply]
> So internally, PyChamber already does this using a `NetworkSet` :D All data and both polarizations are stored in one `NetworkSet` with parameters used to tag each network properly. I just worry about the overhead. Creating a new `Network` for each measurement seems like a lot!

Well, is it really such a big problem ?

Once done, the data can be saved under .hdf, MDIF or similar kind of file (then zipped, etc).

> What would you think about an ndarray with shape `(len(thetas), len(phis), len(freqs)`? 

We should think about as generic as possible, something that could fit many situations: some would like to measure their antenna in a single plane, some in 3D, some in 4D (different time, different voltage, different media, etc) and probably a lot of other special cases, etc... So I'm not in favour to fix the shape of an N-dimensional data structure to store the data. 

The common denominator is that we are talking about:
- frequency measurements (one frequency or many)
- Network parameters (s, or other)
so each configuration fits well the definition of a Network with `param`.

> Also if we _did_ use `NetworkSet`, I think it (or `Network`, not sure) would need some updates. Because params is a dictionary, these objects are _not pickleable_. Additionally, `NetworkSet.write_spreadsheet()` does not work (I've not filed an issue yet, apologies!)

Yes, but pickling should be avoided in general I think. It's better to recommand .hdf, .mdif or other kind of file for that.

**Reply by cafeclimber** [reply]
> Once done, the data can be saved under .hdf, MDIF or similar kind of file (then zipped, etc).

Ah that's my fault. I'd not looked into those formats. With that being the case though, I think it might be worth removing those methods from `Network`/`NetworkSet` as currently it just throws a cryptic error about `dict_keys` being unpicklable, which is not great for the user.

> We should think about as generic as possible, something that could fit many situations: some would like to measure their antenna in a single plane, some in 3D, some in 4D (different time, different voltage, different media, etc) and probably a lot of other special cases, etc... >So I'm not in favour to fix the shape of an N-dimensional data structure to store the data.

Mmm. I hadn't considered measurements in time. I think this goes back to needing to nail down the purpose of such a class. If it's just for storing and visualizing / analyzing pattern data, then I think having just far field mag/phase in a `(thetas,phis,freqs)` shaped `ndarray` would be appropriate. That shape would change depending on the specific extents / resolution on initialization, but would still be 3D (in the case of a single cut plane, the shape would be something like `(thetas, 1, freqs)`). If that's not what we want for the class however, then of course that structure would be different. 

Thinking more about something like 4+D measurements (with varying measurement configuration / time), could we not just suggest using a `list(Antenna)` and tag each `Antenna` with that metadata? I just wonder what the interface would look like as these potential other parameters grow. IMO, keeping the `Antenna` class as simple as possible allows extensibility / subclassing to fit a user's specific needs.

Now of course if we have as a feature of the class, being able to connect other `Networks` to do things like matching, using something like `list(list(list(Network)))` makes more sense, but still keeping the shape 3D as mentioned above.

**Reply by jhillairet** [reply]
> > We should think about as generic as possible, something that could fit many situations: some would like to measure their antenna in a single plane, some in 3D, some in 4D (different time, different voltage, different media, etc) and probably a lot of other special cases, etc... >So I'm not in favour to fix the shape of an N-dimensional data structure to store the data.
> 
> Mmm. I hadn't considered measurements in time. I think this goes back to needing to nail down the purpose of such a class. If it's just for storing and visualizing / analyzing pattern data, then I think having just far field mag/phase in a `(thetas,phis,freqs)` shaped `ndarray` would be appropriate. That shape would change depending on the specific extents / resolution on initialization, but would still be 3D (in the case of a single cut plane, the shape would be something like `(thetas, 1, freqs)`). If that's not what we want for the class however, then of course that structure would be different.

Again, I would recommend not preclude a given data format for the kind of data or the coordinates. For example, during my PhD, I did antenna measurements in close-field, in (x,y) planes for different distances from the antenna, and storing both reflexion and transmission data... 

To me, a `NetworkSet` fits what is needed for storing the data. Then, manipulations are performed using `NetworkSet `or `Antenna `methods, like for [interpolation](https://scikit-rf.readthedocs.io/en/latest/examples/networksets/Interpolating%20Network%20Sets.html), [sorting](https://scikit-rf.readthedocs.io/en/latest/examples/networksets/Sorting%20Network%20Sets.html) or [selecting a given set of parameters](https://scikit-rf.readthedocs.io/en/latest/tutorials/NetworkSet.html#Named-Parameters). It would also work in case of more dimensions (like time, voltage, etc)

**Reply by cafeclimber** [reply]
That's all fine, the main sticking point for me is just whether or not the overhead of creating on the order of e.g. 65,000+ (360x180) `Network` objects is acceptable. 

If so, I'll get started on an implementation. If not though, what are the alternatives.

**Reply by jhillairet** [reply]
> That's all fine, the main sticking point for me is just whether or not the overhead of creating on the order of e.g. 65,000+ (360x180) `Network` objects is acceptable.

That's a good question indeed. I think it should be easy to test, by taking a measured example and create a big NetworkSet by duplicating it a big number of times to see how the memory grows... ;)

**Reply by cafeclimber** [reply]
Did a quick test. (lmk if this methodology is nonsense).

I used this code (Halo just provides a spinner to monitor progress

```python
from pympler import asizeof
import matplotlib.pyplot as plt

from halo import Halo

import skrf

def using_copy():
    sizes = []
    ntwk = skrf.data.ring_slot_meas
    print(f"Single Network: {asizeof.asizeof(ntwk)}")
    ntwk_set = skrf.NetworkSet([ntwk])
    print(f"Single Network NetworkSet: {asizeof.asizeof(ntwk_set)}")

    with Halo(text="n=", spinner='dots') as spinner:
        for n in range(1, 100_000, 1000):
            spinner.text = f"n={n}"
            ntwks = [ntwk.copy() for i in range(n)]
            ntwk_set_tmp = skrf.NetworkSet(ntwks)
            sizes.append(asizeof.asizeof(ntwk_set_tmp))

    plt.plot(sizes)
    plt.show()

    print(f"Size of 100000 ntwk NetworkSet: {asizeof.asizeof(ntwk_set_tmp)}")


if __name__ == "__main__":
    using_copy()
```

with these results

```console
Single Network: 6256
Single Network NetworkSet: 6776
Size of 100000 ntwk NetworkSet: 527487808
```
![ntwk_set_from_copies](https://user-images.githubusercontent.com/10188900/191765921-4a5ab1a2-9176-4d34-84c8-68e859c75d60.png)

100,000 ntwks would be roughly two time snapshots of full spherical (1 deg increments) measurements. So for two time snapshots, that's ~500MiB. I don't know how much that data would change by moving away from `NetworkSet`

**Comment by FranzForstmayr** [comment]
I'm not sure if it's the best way to store every phi/theta combination in a separate 1-port network. Wouldn't it be more straight forward to use a spherical coordinate system in combination with the frequency as data structure (xarray, or even pandas). The index would be a combination of frequency, theta and phi. I  think it would be easier when doing tasks like interpolation between frequency points or theta/phi combinations.

It would be great to create a 1-port network of given phi/theta obviously.

**Reply by cafeclimber** [reply]
Right, I keep going back to just using a 3D numpy array that's shape (theta,phi, frequencies) and only store far-field data. 

Then we could implement indexing like you say and if we're only interested in say, gain at bore sight over frequency, we could just convert to a `Network` on the fly. Eg.

```python
ant = Antenna.dipole("2.45GHz")
ntwk = ant[theta][phi]
type(ntwk) # skrf.Network
```

**Reply by jhillairet** [reply]
> It would be great to create a 1-port network of given phi/theta obviously.

I think we should imagine antenna with more than 1-port. For example, phased arrays have more than one port.

**Comment by cafeclimber** [comment]
Starting work on this implementation, and here are some thoughts I've had:

- If we want to do things like connecting a matching network to an antenna, I feel like that is fundamentally different from something like field measurements and I don't know how we can combine the two in a `NetworkSet`. I think we could do something like having a 3D representation of fields as a 3D numpy array with customizable endpoints and resolution as well as having a single N-port network to represent the input / isolation characteristics of multi-port antennas. If we want to look at input / isolation across e.g. voltage, temperature, etc. we could have a `NetworkSet` to represent connection to the antenna, but still have a 3D array for fields.
- For plotting, do we want a signature like:
```python
    def plot_az_cut(
        self,
        theta: float = 90.0,
        projection: str = "polar",
        db: bool = True,
        normalized: bool = True,
    ) -> None:
        ...
```
I know there is the `plotting` module which contains `plot_polar` but I don't know how to pass things like `set_theta_zero_location` to those plots.

**Reply by jhillairet** [reply]
Here is how I see things, writing as thinking, and comments are very welcome.

* an antenna is intrinsically an N-port device (defined in a circuit by its network parameters vs RF frequency), with additional information on top a classic N-port, like 3D fields information or other. 
  * Hence, an `Antenna` class seems to naturally inherits from `Network`
  * but the question of the storage of the 3D fields remains. Is there a convention used in the antenna measurement community to store data? (I don't really work with "real antenna", the ones that radiate, so I don't know)
* A measurement/simulation of an antenna consists in measuring/simulating its 3D fields and its network parameters for a given configuration (environment, electronic setup, etc).

* If one assumes that the network parameters are not affected by the measurement of the 3D fields (ie. if S11 is the same when the field measurement is made for any angle), then there is no need so store the network parameter for each measurement point. So if correct, I change my mind: there is no need of a list of `Network` (aka `NetworkSet). A 3D dataset associated to a Network would be sufficient (and TBD).

So question for experts: During the measurement of the 3D fields of antenna, are the network parameters also saved against frequency? Depending of this that would change the way to organize the things.

**Reply by cafeclimber** [reply]
For me when I'm taking measurements in the chamber, our measurements are far-field and across frequency. This is why I used a `NetworkSet` in PyChamber, but the difference there is I'm not looking at any input characteristics when taking those measurements. We usually do those on a VNA outside the chamber to calibrate / minimize cable runs / etc. If I were to do it again, I think I would instead do what we're describing here: a 3D numpy array of shape `(theta,phi,freq)`. For our purposes, I think that makes sense for far field in addition to having a `NetworkSet` for input characteristics. Now if users want to have antennas with differing configurations, we could have a those fields be stored like

```python
fields = [
    {
        'voltage': 1V,
        'temp': 20,
        'other metadata': ...,
        'data': # field measurements as numpy array
    }
]
```
or maybe even just do that and for each element of the list have the metadata, the field measurements, _and_ the `NetworkSet` characterizing the input. Then we could do `sel` similarly to how it's done in 'NetworkSet`. Now if that's the best interface? I leave to other folks haha.

I do wonder if we should consider a way for users with a phased array or other N-port antenna to specify the complex weights applied to the ports during measurement. Maybe there's already a solution.

**Reply by cafeclimber** [reply]
As to how those measurements are stored? I don't know if there's a standard. For the software that came with the manufacturer of our antenna positioner (Diamond Engineering), they use their own text-based format for storing files with the ability to export to Excel. Now those files are only one per measurement which presents a potential improvement for us. Having a way to store all that data in a single object I think greatly simplifies post-processing. Right now, we have to keep track of / tag all our measurements in their filenames with e.g. `(Sept28th/array_30deg.dat)` which is awful.

My advisor had suggested using [Remcom's UAN file format](https://support.remcom.com/xfdtd/reference/results/uan-file-format.html) for PyChamber, but this `Antenna` discussion happened to come up at the same time so I chose to focus on it instead. I don't know if that format could be built on for writing data to disk or if we could instead write our own serializer / file format ([insert xkcd joke](https://xkcd.com/927/))
